include/envoy/common/case_insensitive.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <string>
+
+#include "common/common/assert.h"
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+
+/**
+ * Wrapper for case insensitive string to generally avoid needless case insensitive compares. The
+ * wrapper will uniformly convert the string content to lowercase.
+ */
+class LowerCaseStrBase {
+public:
+  explicit LowerCaseStrBase(absl::string_view new_string) : string_(new_string) { lower(); }
+
+  LowerCaseStrBase(LowerCaseStrBase&&) = default;
+  LowerCaseStrBase& operator=(LowerCaseStrBase&&) = default;
+
+  LowerCaseStrBase(const LowerCaseStrBase&) = default;
+  LowerCaseStrBase& operator=(const LowerCaseStrBase&) = default;
+
+  const std::string& get() const { return string_; }
+  std::string& get() { return string_; }
+
+  bool operator==(const LowerCaseStrBase& rhs) const { return string_ == rhs.string_; }
+  bool operator!=(const LowerCaseStrBase& rhs) const { return string_ != rhs.string_; }
+  bool operator<(const LowerCaseStrBase& rhs) const { return string_.compare(rhs.string_) < 0; }
+
+  friend std::ostream& operator<<(std::ostream& os, const LowerCaseStrBase& lower_case_string) {
+    return os << lower_case_string.string_;
+  }
+
+protected:
+  void lower() {
+    std::transform(string_.begin(), string_.end(), string_.begin(), absl::ascii_tolower);
+  }
+
+  std::string string_;
+};
+
+template <bool (*V)(absl::string_view)> class ValidatedLowerCaseStr : public LowerCaseStrBase {
```

Comment:
I'm not sure why we need this to be a template.

Reply:
@lizan Because I want to get a protocol-independent lower case string and be able to set a different validator. The template is used so that the validator is pluggable.

Here are some discussions about this question in original PR #16049. 
@jmarantz  https://github.com/envoyproxy/envoy/pull/16049#discussion_r623008023

Reply:
OK I see, then this should be commented why it is template. and in Http, it should be named like validateHttpLowerCaseString.

Reply:
TBH my recollection from the.other PR is that there are a few jumbled concepts here and I want to just quickly mention my line of thinking.

IMO if you really need a validator,  forget whether there's a compelling reason to prefer
 * template
 * virtual method
 * function-pointer
But if we do use a function-pointer then in IMO it could be a std::function rather than a templatized function pointer.

Also I think once we are plugging in validation, there's no compelling reason to use LowerCaseString just for Tracing's sake, unless there's something about Tracing that wants strings to be lower-cased. IDK whether that's the case.

That said, I have no fundamental objection to moving LowerCaseString to common as long as it doesn't make anything slower for HTTP headers, which are performance critical.

I don't understand why LowerCaseString has to be paired with a validator though. Can't you just leave LowerCaseString exactly like it is, and move it out of http into common? That seems like a legitimate concept on its own. You could make another layer that adds validation if this is helpful in some way.

Reply:
@lizan I will add some new comments to explain why we need this template.
In Http, I kept the original Http::LowerCaseString name so as not to affect the existing code as much as possible. After all, the name Http::LowerCaseString is already widely used.

Reply:
So this isn't a move...it's a copy?

There's nothing inherently wrong with doing a move and changing all the references though that will make the PR large, so you could always do the move of the code and leave a forwarding reference in Http, then follow up with a cleanup.

But I don't understand why the Trace functionality in the other PR needs lower-casing at all.

Reply:
@jmarantz 
- First, about the performance: This modification will not affect the performance of Http::LowerCaseString since there is no new logic. The introduction of templates may increase the compilation overhead slightly.
- Second, about the relationship between validator and LowerCaseString: I provide a class `LowerCaseStrBase` as a common implementation of LowerCaseString which with no any validation and can be used directly. Then I provide `ValidatedLowerCaseStr` template just as a helper for who wants to add validation to LowerCaseString, for example, Http. 
So that means I've got two layers as you think. One layer is the plain `LowerCaseString`, and the second layer is the `ValidatedLowerCaseStr` that can mount a different validator.

Reply:
@jmarantz At the source code level, it is move.
I mean that in Http I kept the previous name `LowerCaseString` instead of renaming it to `ValidatedHttpLowerCaseString` or something like that to make PR clean.

Reply:
OK...sorry, TBH I'm skeptical about the overall strategy and also concerned there may be overhead added to http headers.

But for the moment maybe it'd be better to focus on strategy. Do you want to describe the high level of what you are trying to achieve in a bug or a google doc? It might be easier to iterate there.

Reply:
Thanks for you comments and patience. @jmarantz 

In the original PR #16049, there is the high level design of the general tracing. And I remember I had a more detailed discussion with you in slack earlier.

For now, I think we can try to focus on this PR first. Because this PR don't involve the design of the new general tracing yet, just some pre-work. This is also the reason why I split the original PR into multiple PRs.
I think we can discuss the design of general tracing in detail in the new PR (One or two weeks after the completion of this PR). 

For this PR I think the most important points are the following:
* whether the PR is clean or not.
* whether the PR is able to keep the existing functionality working.
* whether the PR can guarantee that the existing performance is not affected.
* Whether the PR achieves its purpose: to strip LowerCaseString from Http and get a general LowerCaseString.